planner/core/task.go

```diff
@@ -510,15 +764,55 @@ func (p *PhysicalHashAgg) attach2Task(tasks ...task) task {
 				partialAgg.SetChildren(cop.indexPlan)
 				cop.indexPlan = partialAgg
 			}
+			cop.addCost(p.GetCost(inputRows, false))
 		}
+		// In `newPartialAggregate`, we are using stats of final aggregation as stats
+		// of `partialAgg`, so the network cost of transferring result rows of `partialAgg`
+		// to TiDB is normally under-estimated for hash aggregation, since the group-by
+		// column may be independent of the column used for region distribution, so a closer
+		// estimation of network cost for hash aggregation may multiply the number of
+		// regions involved in the `partialAgg`, which is unknown however.
 		t = finishCopTask(p.ctx, cop)
+		inputRows = t.count()
 		attachPlan2Task(finalAgg, t)
 	} else {
 		attachPlan2Task(p, t)
 	}
-	t.addCost(t.count()*cpuFactor*hashAggFactor + cardinality*createAggCtxFactor)
-	if p.hasDistinctFunc() {
-		t.addCost(t.count() * cpuFactor * distinctAggFactor)
-	}
+	// We may have 3-phase hash aggregation actually, strictly speaking, we'd better
+	// calculate cost of each phase and sum the results up, but in fact we don't have
+	// region level table stats, and the concurrency of the `partialAgg`,
+	// i.e, max(number_of_regions, DistSQLScanConcurrency) is unknown either, so it is hard
+	// to compute costs separately. We ignore region level parallelism for both hash
+	// aggregation and stream aggregation when calculating cost, though this would lead to inaccuracy,
+	// hopefully this inaccuracy would be imposed on both aggregation implementations,
+	// so they are still comparable horizontally.
+	// Also, we use the stats of `partialAgg` as the input of cost computing for TiDB layer
+	// hash aggregation, it would cause under-estimation as the reason mentioned in comment above.
+	// To make it simple, we also treat 2-phase parallel hash aggregation in TiDB layer as
+	// 1-phase when computing cost.
+	t.addCost(p.GetCost(inputRows, true))
 	return t
 }
+
+// GetCost computes the cost of hash aggregation considering CPU/memory.
+func (p *PhysicalHashAgg) GetCost(inputRows float64, isRoot bool) float64 {
+	cardinality := p.statsInfo().RowCount
+	numDistinctFunc := p.numDistinctFunc()
+	var cpuCost float64
+	if isRoot {
+		cpuCost = inputRows * cpuFactor * float64(len(p.AggFuncs))
+		divisor, con := p.cpuCostDivisor(numDistinctFunc > 0)
+		if divisor > 0 {
+			cpuCost /= divisor
+			// Cost of additional goroutines.
+			cpuCost += (con + 1) * concurrencyFactor
```

Comment:
`divisor == 0` should also consider cost of addtional goroutines?

Reply:
If `divisor == 0`, the hash aggregation is executed not in parallel, so we don't need to add the cost of worker goroutines?

Reply:
`divisor == 0` represents `concurrency == 1`. It also builds goroutines in hash aggregation

Reply:
According to https://github.com/pingcap/tidb/blob/268cbf1a5d2491b4242f26109e05fbdf684e16fa/executor/builder.go#L1039, if both concurrency is 1, it would not execute in parallel.

Reply:
OK.